arch/risc-v/src/common/riscv_percpu.c

```diff
@@ -0,0 +1,117 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/irq.h>
+
+#include <arch/barriers.h>
+
+#include <assert.h>
+#include <stdint.h>
+
+#include "riscv_internal.h"
+#include "riscv_percpu.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define HART_CNT    (CONFIG_ARCH_CPU_COUNT)
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct riscv_percpu_s g_scratch[HART_CNT];
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: riscv_percpu_init
+ *
+ * Description:
+ *   Initialize the per CPU structures, should only be done on the boot
+ *   hart.
+ *
+ ****************************************************************************/
+
+void riscv_percpu_init(void)
+{
+  int i;
+
+  for (i = 0; i < HART_CNT; i++)
+    {
+      g_scratch[i].hartid = i;
+    }
+}
+
+/****************************************************************************
+ * Name: riscv_percpu_get_addr
+ *
+ * Description:
+ *   Get add a hart to the per CPU area
+ *
+ * Input Parameters:
+ *   hartid - Hart number
+ *
+ ****************************************************************************/
+
+void riscv_percpu_add_hart(uintptr_t hartid)
+{
+  /* Hart IDs go from 0...4 */
+
+  DEBUGASSERT(hartid < HART_CNT);
+
+  /* Set the scratch register value to point to the scratch area */
+
+  WRITE_CSR(sscratch, &g_scratch[hartid]);
+
+  /* Make sure it sticks */
+
+  __DMB();
+}
+
+/****************************************************************************
+ * Name: riscv_percpu_get_hartid
+ *
+ * Description:
+ *   Get harts own hartid by reading it from the per CPU area. This is safe
+ *   to use from lower privilege modes (than M-mode).
+ *
+ * Returned Value:
+ *   Hart id
+ *
+ ****************************************************************************/
+
+uintptr_t riscv_percpu_get_hartid(void)
+{
+  uintptr_t scratch = READ_CSR(sscratch);
```

Comment:
Do we need `sscratch` vs `mscratch` selection here?

Reply:
The intent was to use sscratch only for the percpu structure at least for now. 

Asking for hartid in m-mode is simple, just `READ_CSR(mhartid); `